app/src/main/java/io/apicurio/registry/util/ArtifactTypeUtil.java

```diff
@@ -84,20 +108,41 @@ public static ArtifactType discoverType(ContentHandle content, String contentTyp
         } catch (Exception e) {
             // Apparently it's not JSON.
         }
-        
+
         // Try protobuf (only if we haven't already)
         if (!triedProto) {
             ArtifactType type = tryProto(content);
             if (type != null) {
                 return type;
             }
         }
-        
+
         // Try GraphQL (SDL)
         if (tryGraphQL(content)) {
             return ArtifactType.GRAPHQL;
         }
-        
+
+        // Try the various XML formatted types
+        try (InputStream stream = content.stream()) {
+            Document xmlDocument = threadLocaldocBuilder.get().parse(stream);
+            Element root = xmlDocument.getDocumentElement();
+            String ns = root.getNamespaceURI();
+
+            // XSD
+            if (ns != null && ns.equals("http://www.w3.org/2001/XMLSchema")) {
+                return ArtifactType.XSD;
+            } // WSDL
+            else if (ns != null && (ns.equals("http://schemas.xmlsoap.org/wsdl/")
+                    || ns.equals("http://www.w3.org/ns/wsdl/"))) {
+                return ArtifactType.WSDL;
+            } else {
+                // default to XML since its been parsed
+                return ArtifactType.XML;
+            }
```

Comment:
Now I see better why you had the null check up top.  But this is good I think.

Reply:
👍